### PR TITLE
Sync connected devices management

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1204400287298647/f 

### Description
Adds functionality around connected (local/remote) device management. 
- rename local device
- remove remote device
- list all connected devices

### Steps to test this PR

1️⃣ Remove `FakeDeviceSyncState` class

_list connected devices_
- [ ] From one device, go to sync settings and create an account.
- [ ] From a second unauthenticated device, go to sync settings
- [ ] Now join the existing account (toggle on -> recover your synced data -> read the QR code)
- [ ] When device joins the account, go back to sync main screen
- [ ] you should see both device listed there
- [ ] everytime the user opens the scree, we will fetch the connected device list (loading shimmer will appear)

_rename device_
- [ ] In sync settings, with an authenticated device
- [ ] click on the 3 dots menu for "this device"
- [ ] rename the device
- [ ] if the operation succeeds, the new name will be applied
- [ ] if the operation fails, name shouldn't be updated
- [ ] on another connected device, you can access the sync settings to see the change reflected there

_remove device_
- [ ] In sync settings, with an authenticated device
- [ ] you can remove a remote device (you should have 2 connected devices)
- [ ] click on the 3 dots menu for the remote device
- [ ] click on remove device
- [ ] if the operation succeeds, the device will be removed from the list
- [ ] if the operation fails, the device will stay
- [ ] shimmer should indicate progress
- [ ] the other device is going to be on a weird state until we implement https://app.asana.com/0/0/1204341643230613/f

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3098** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3098" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3098?utm_source=stack-comment).

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt [hunk 1]
```diff
@@ -50,98 +54,186 @@ class SyncActivityViewModel @Inject constructor(
...
         viewModelScope.launch {
-            updateViewState()
+            updateAccountDetails()
         }
     }
 
     fun onToggleClicked(isChecked: Boolean) {
         viewModelScope.launch {
-            viewState.emit(viewState.value.copy(isDeviceSyncEnabled = isChecked))
+            viewState.emit(viewState.value.toggle(isChecked))
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt:86]: 💯

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt [hunk 2]
```diff
@@ -50,98 +54,186 @@ class SyncActivityViewModel @Inject constructor(
...
-                    updateViewState()
-                }
-                is Success -> {
-                    Timber.i("deleteAccount success")
-                    updateViewState()
-                }
-            }
+            viewState.emit(viewState.value.hideAccount())
+            syncRepository.deleteAccount()
+            hideOrShowAccountDetails()
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt:135]: why are you emitting two view states after each other? is this really necessary?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt [hunk 3]
```diff
@@ -50,98 +54,186 @@ class SyncActivityViewModel @Inject constructor(
...
+            viewModelScope.launch(dispatchers.io()) {
+                fetchRemoteDevices()
+            }
+        }
+    }
+
+    private suspend fun hideOrShowAccountDetails() {
+        if (!syncRepository.isSignedIn()) {
+            viewState.emit(signedOutState())
+            return
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt:215]: is this necessary?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt [hunk 4]
```diff
@@ -50,98 +54,186 @@ class SyncActivityViewModel @Inject constructor(
...
-                is Error -> {
-                    updateViewState()
-                }
-                is Success -> {
-                    updateViewState()
-                }
-            }
+            viewState.emit(viewState.value.hideAccount())
+            syncRepository.logout(connectedDevice.deviceId)
+            hideOrShowAccountDetails()
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt:114]: ditto

**Review by malmstein — COMMENTED** [review]
Added comments after a first pass, will go through test steps tomorrow.

**Review by malmstein — APPROVED** [review]
reviewed offline

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 5]
```diff
@@ -263,7 +270,8 @@ class AppSyncRepository @Inject constructor(
         return syncStore.token ?: ""
     }
 
-    override fun getThisConnectedDevice(): ConnectedDevice {
+    override fun getThisConnectedDevice(): ConnectedDevice? {
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:279]: As commented, we will do that when reviewing the SyncRepository api interface https://app.asana.com/0/0/1204139091554737/f

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt (hunk 2 above)
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt]: here we emit hideAccount for update the UI without waiting for the backend response. If it fails, the UI will show the account again. Just an optimistic approach

**Pull request merged**